File: 11. Container With Most Water/main.js
/**
 * @param {number[]} height
 * @return {number}
 */
// Brute force solution
// Time complexity: O(n^2)
// Space complexity: O(1)
var maxArea = function (height) {
  let maxArea = 0;
  const n = height.length;

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const area = Math.min(height[i], height[j]) * (j - i);
      maxArea = Math.max(maxArea, area);
    }
  }

  return maxArea;
};

// Two-pointer solution
// Time complexity: O(n)
// Space complexity: O(1)
var maxAreaTwoPointer = function (height) {
  let left = 0;
  let right = height.length - 1;
  let maxArea = 0;
  while (left < right) {
    const area = Math.min(height[left], height[right]) * (right - left);
    maxArea = Math.max(maxArea, area);
    if (height[left] < height[right]) {
      left++;
    } else {
      right--;
    }
  }
  return maxArea;
};

File: 11. Container With Most Water/main.py
from typing import List
class Solution:
    def maxArea(self, height: List[int]) -> int:
        max_water = 0
        left = 0
        right = len(height) - 1

        while left <= right:
            max_water = max((right - left)*min(height[left],height[right]),max_water)
            if height[left] < height[right]:
                left+=1
            else:
                right-=1

        return max_water

if __name__ == "__main__":
    s = Solution()
    s.maxArea([1,8,6,2,5,4,8,3,7])

    print("🚀 ------------------------------------------------------------------------------🚀")
    print("🚀 ~ main.py:14 ~ s.maxArea([1,8,6,2,5,4,8,3,7]):", s.maxArea([1,8,6,2,5,4,8,3,7]))
    print("🚀 ------------------------------------------------------------------------------🚀")

